Description: Allow overriding plugin install directory
Author: Jeremy Bicha <jbicha@ubuntu.com>
Index: burner-3.12.1/configure.ac
===================================================================
--- burner-3.12.1.orig/configure.ac
+++ burner-3.12.1/configure.ac
@@ -295,7 +295,7 @@ AM_CONDITIONAL(BUILD_NAUTILUS, test x"$b
 # Installation path for plugins
 #
 
-BURNER_PLUGIN_DIRECTORY=${libdir}/burner3/plugins
+BURNER_PLUGIN_DIRECTORY=${BURNER_PLUGIN_DIRECTORY:-'${libdir}/burner3/plugins'}
 AC_SUBST(BURNER_PLUGIN_DIRECTORY)
 
 dnl ****************check for libburn (optional)**************
